### PR TITLE
Second pass at dispatcher / router

### PR DESCRIPTION
**Pull request by tswicegood**
## Why are we doing this?

This PR adds a `dispatcher` and a `router` and moves message routing into to the latter. The idea is here is an incremental step toward having all communications between the various concerns handled via events.

There are a few motivations driving this change:

* The `src/app.js` conflates several concerns (Express routing, session management, message routing, public API for emitting events)
* Handling everything via a `dispatcher` allows third-party developers to hook into the various parts of the system by monitoring events that are triggered
* Easier testing – for example the test for the express route isn't that the various messages are routed correctly. Instead, you test for the route to emit an event and that a handler for that event is registered by default. Once that's in place you test the handler independently.
* The three preceding points can be summed up as "decoupling."

## Did you document your work?

* [x] Session documentation needs to be updated to show how to save it

## How can someone test these changes?

Steps to manually verify the change:

1. `npm install`
2. `npm test`
3. Verify that the various `router` specifications are present (current 44 passing, 2 pending)

## What possible risks or adverse effects are there?

* [x] ~~~💣  session saving is currently broken~~~
* Code churn, but that can be a good thing.
* Moving to a fully event-driven system takes a little getting used to (case in point: Mocha internals)

## What are the follow-up tasks?

* Sessions saving needs to be modified because it is happening async now
* `Messenger` needs to be extract into an object by itself
* `app.Messenger` needs to be refactored into a simple Express app that emits events everywhere
* Session code needs to be extracted into its own module / event handler

None of these are necessary for merge and are more next steps.

### CODE REVIEW

File: src/app.js [hunk 1]
```diff
@@ -44,6 +48,11 @@ class Messenger extends EventEmitter {
     }
     this.options.emitGreetings = !!emitGreetings;
 
+    router.init(this);
+    dispatcher.on('session.changed', ({session}) => {
```
Comment (tswicegood) [inline src/app.js]: Curious on your thoughts, @crccheck & @stripethree. My thought is to allow an object to be passed and only lay claim to `session` as a property so you could emit other things on `session.changed`. Not sure if it's needed or not so I'm kinda on the fence about this verse `… (session) => {`.

**Comment by tswicegood** [comment]
@crccheck @stripethree this is ready for a pass from y'all. There's still more to do, but I think this covers the basics. Separating `Messenger` from an `App` object and extract all of the session stuff into it's own object can be tackled separately I think.

File: src/app.js [hunk 2]
```diff
@@ -25,6 +27,8 @@ const internals = {};
 const DEFAULT_GREETINGS_REGEX = /^(get started|good(morning|afternoon)|hello|hey|hi|hola|what's up)/i;
 const DEFAULT_HELP_REGEX = /^help\b/i;
 
+// TODO Extract this class into `index` and refactor `App` to be only the
+//      Express configuration and set up.
```
Comment (tswicegood) [inline src/app.js]: Is that not the convention? I'm trying to line up underneath the TODO above to show that they're one thing.

File: src/app.js [hunk 3]
```diff
@@ -115,50 +125,49 @@ class Messenger extends EventEmitter {
...
+    return (session = defaultSession) => {
+      // WISHLIST: logic to handle any thundering herd issues: https://en.wikipedia.org/wiki/Thundering_herd_problem
+      if (session.profile) {
+        return session;
+      } else if (messagingEvent.sender.id === config.get('facebook.pageId')) {
+        // The page does not have a public profile and calling the Graph API here will always yield a 400.
+        session.profile = {};
+        return session;
+      } else {
+        return this.getPublicProfile(messagingEvent.sender.id)
```
Comment (tswicegood) [inline src/app.js]: Not sure I follow, @stripethree. This was an extraction of the code that was already in the promise chain for dealing with sessions before emitting the event. I started to pull all of these out into their own module to decouple them from `App`, but wanted to stop short least this turn into a huge refactor.

Comment (tswicegood) [inline src/app.js]: Actually, now I finally get it. You're talking about pulling `getPublicProfile` into this function, right? I'm ¯\_(ツ)_/¯ on that. I think there's still some extraction to be done here so we could remove that method or leave it for now knowing this all still needs some tweaking.

Comment (stripethree) [inline src/app.js]: Yes, I should have caveated my comment as it was definitely me thinking through the keyboard.

File: src/app.js (hunk 2 above)
Comment (stripethree) [inline src/app.js]: :banghead: For some reason, I read it as a separate comment, not a continuation of the previous line. My bad, carry on.

**Comment by tswicegood** [comment]
Minus the conflicts that need to be resolved, this is ready to merge. How would you like me to handle merge conflicts (my normal `git merge` workflow won't work here).

File: src/app.js [hunk 4]
```diff
@@ -15,6 +15,8 @@ const urlJoin = require('url-join');
 
 const config = require('./config');
 const conversationLogger = require('./conversationLogger');
+const dispatcher = require('./dispatcher');
```
Comment (crccheck) [inline src/app.js]: Why is there a dispatcher _and_ a router? This seems like something that will cause a lot of confusion.

Comment (tswicegood) [inline src/app.js]: @crccheck the idea is to have `dispatcher` be the event bus between various separations of concerns. This is a first step toward that. The next is pulling the Express code into its own code and the session code out into its own module. This the first step toward that.

**Comment by tswicegood** [comment]
> Yes, this is splitting code up into smaller modules, but they're still tied together so you can't understand what's going on unless you have all the files open in front of you.

See adverse side effects:

> Moving to a fully event-driven system takes a little getting used to (case in point: Mocha internals)

It definitely changes up the paradigm. I can make a case for fully event-driven internals _or_ for keeping it as-is where everything is more defined. Reading through `src/app.js` for these changes left me with a sense that the `Messenger` class was doing too much.

That said, you and @stripethree have much more familiarity with this code and its uses so if you don't think it makes sense, I'm happy to close it w/o merge.

**Comment by stripethree** [comment]
Follow-up on one of my previous comments:
> I can create a public repo with an example and instructions.

I went ahead and did that, [here].

[here]: https://github.com/stripethree/enterprise-fbm

**Comment by tswicegood** [comment]
> Lastly, I would like to request an update to the Why for this PR to cover more of how this benefits the SDK, esp. in conjunction with the follow-ons.

@stripethree updated the description. Sorry, missed that earlier.

**Comment by stripethree** [comment]
IRL convo w/@tswicegood: closing this out to have a clean view on 1.0.0

**Comment by crccheck** [comment]
I read through this PR once a day, but I don't have anything new to add to the conversation.

**Pull request closed without merging**